pkg/kubectl/apis/manifest/v1alpha1/types.go

```diff
@@ -0,0 +1,138 @@
+	ObjectLabels map[string]string `json:"objectLabels,omitempty" yaml:"objectLabels,omitempty"`
+
+	// Annotations to add to all objects.
+	ObjectAnnotations map[string]string `json:"objectAnnotations,omitempty" yaml:"objectAnnotations,omitempty"`
+
+	// A Base entry specifies the relative paths within the package.
+	// It could be any format that kubectl -f allows, i.e. files, directories, URLs and globs.
+	Bases []string `json:"bases,omitempty" yaml:"bases,omitempty"`
+
+	// An Overlay entry is very similar to an Base entry.
+	// It specifies the relative paths within the package, and could be any format that kubectl -f allows.
+	// It should be able to be merged by Strategic Merge Patch on top of its corresponding Base if any.
+	Overlays []string `json:"overlays,omitempty" yaml:"overlays,omitempty"`
+
+	// List of configmaps to generate from configuration sources.
+	// Base/overlay concept doesn't apply to this field.
+	// If a configmap want to have a base and an overlay, it should go to Bases and Overlays fields.
+	Configmaps []ConfigMap `json:"configmaps,omitempty" yaml:"configmaps,omitempty"`
+
+	// List of secrets to generate from secret sources.
+	// Base/overlay concept doesn't apply to this field.
+	// If a secret want to have a base and an overlay, it should go to Bases and Overlays fields.
+	Secrets []Secret `json:"secrets,omitempty" yaml:"secrets,omitempty"`
+
+	// Whether PersistentVolumeClaims should be deleted with the other resources.
+	OwnPersistentVolumeClaims bool `json:"ownPersistentVolumeClaims,omitempty" yaml:"ownPersistentVolumeClaims,omitempty"`
+
+	// Whether recursive look for Kube-manifest.yaml, similar to `kubectl --recursive` behavior.
+	Recursive bool `json:"recursive,omitempty" yaml:"recursive,omitempty"`
+
+	// Whether prune resources not defined in Kube-manifest.yaml, similar to `kubectl apply --prune` behavior.
+	Prune bool `json:"prune,omitempty" yaml:"prune,omitempty"`
+}
+
+// ConfigMap contains the metadata of how to generate a configmap.
+type ConfigMap struct {
+	// The type of the configmap. e.g. `env` and `file`.
+	Type string `json:"type,omitempty" yaml:"type,omitempty"`
+
+	// Name prefix of the configmap.
+	// The full name should be Manifest.NamePrefix + Configmap.NamePrefix + hash(content of configmap).
+	NamePrefix string `json:"namePrefix,omitempty" yaml:"namePrefix,omitempty"`
+
+	// Configuration source file name.
+	File string `json:"file,omitempty" yaml:"file,omitempty"`
+}
+
+// Secret contains the metadata of how to generate a secret.
+type Secret struct {
+	// The type of the secret. e.g. `tls`.
+	Type string `json:"type,omitempty" yaml:"type,omitempty"`
+
+	// Name prefix of the secret.
+	// The full name should be Manifest.NamePrefix + Secret.NamePrefix + hash(content of secret).
+	NamePrefix string `json:"namePrefix,omitempty" yaml:"namePrefix,omitempty"`
+
+	// Cert file for the secret.
+	CertFile string `json:"certFile,omitempty" yaml:"certFile,omitempty"`
```

Comment:
What is certfile/keyfile?  Is this assuming the secret is type=tls?  What about other types?

Reply:
Yes, other types will need to be supported. This is an example at the moment.

Reply:
We might as well flesh this out more while we figure out where this file should live.

@mengqiy PTAL:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create_secret.go
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create_configmap.go

and suggest a schema that will accommodate the options supported by the imperative commands.

Reply:
It sounds like the schema should be a superset of our kubectl configmap|secret generator with some additional fields.